message/router/middleware/duplicator.go

```diff
@@ -0,0 +1,36 @@
+package middleware
+
+import (
+	"errors"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+var (
+	notIdempotentError = errors.New("handler is not idempotent")
+)
+
+// Duplicator sends message twice.
+func Duplicator(h message.HandlerFunc) message.HandlerFunc {
+	return func(msg *message.Message) ([]*message.Message, error) {
+		firstProducedMessages, firstErr := h(msg)
+		secondProducedMessages, secondErr := h(msg)
+		allErrorsNil := firstErr == nil && secondErr == nil
+		allErrorsEqual := firstErr != nil && secondErr != nil && firstErr.Error() == secondErr.Error()
+
+		if allErrorsNil || allErrorsEqual {
```

Comment:
IMO we should keep simple responsibility here and just do duplication without checking returned messages and if erros are equal because messages may be a bit different, but still different after each call (for example: different correlation id's, uuids etc.)

We should also probably don't care about checking errors too much. IMO the best approach will be:
- if firstErr != nil, return (it anyway fails, so there is probably no point to try to fail it with duplication)
- if secondErr != nil - just return, something else should handle it

does it make sense? ;) @0michalsokolowski0

Reply:
@roblaszczak You are right. Implementation changed in 105b2bbb1566a1cec843ffc14cf55c3b2654dc3a.

Reply:
much better now :medal_military: